src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -489,6 +516,98 @@ RouterModifyTaskForShardInterval(Query *originalQuery, ShardInterval *shardInter
 
 
 /*
+ * HashedShardIntervalOpExpressions returns a list of OpExprs with exactly two
+ * items in it. The list consists of shard interval ranges with hashed columns
+ * such as (hashColumn >= shardMinValue) and (hashedColumn <= shardMaxValue).
+ *
+ * The function errors out if the given shard interval does not belong to a hash
+ * distributed table.
+ */
+static List *
+HashedShardIntervalOpExpressions(ShardInterval *shardInterval)
+{
+	List *operatorExpressions = NIL;
+
+	Var *hashedGEColumn = NULL;
+	OpExpr *hashedGEOpExpr = NULL;
+	Datum shardMinValue = shardInterval->minValue;
+
+	Var *hashedLEColumn = NULL;
+	OpExpr *hashedLEOpExpr = NULL;
+	Datum shardMaxValue = shardInterval->maxValue;
+
+	Oid integer4GEoperatorId = InvalidOid;
+	Oid integer4LEoperatorId = InvalidOid;
+
+	if (PartitionMethod(shardInterval->relationId) != DISTRIBUTE_BY_HASH)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("cannot created shard interval operator expression for "
+							   "distributed relations other than hash distributed "
+							   "relations")));
+	}
+
+	/* get the integer >=, <= operators from the catalog */
+	integer4GEoperatorId = get_opfamily_member(INTEGER_BTREE_FAM_OID, INT4OID,
+											   INT4OID,
+											   BTGreaterEqualStrategyNumber);
+	integer4LEoperatorId = get_opfamily_member(INTEGER_BTREE_FAM_OID, INT4OID,
+											   INT4OID,
+											   BTLessEqualStrategyNumber);
+
+	/* generate hashed columns */
+	hashedGEColumn = MakeInt4Column();
+	hashedLEColumn = MakeInt4Column();
+
+	/* generate the necessary operators */
+	hashedGEOpExpr = (OpExpr *) make_opclause(integer4GEoperatorId,
+											  InvalidOid, false,
+											  (Expr *) hashedGEColumn,
+											  (Expr *) MakeInt4Constant(
+												  shardMinValue),
+											  InvalidOid, InvalidOid);
+
+	hashedLEOpExpr = (OpExpr *) make_opclause(integer4LEoperatorId,
+											  InvalidOid, false,
+											  (Expr *) hashedLEColumn,
+											  (Expr *) MakeInt4Constant(
+												  shardMaxValue),
+											  InvalidOid, InvalidOid);
+
+	/* update the operators with correct operator numbers and function ids */
+	hashedGEOpExpr->opfuncid = get_opcode(hashedGEOpExpr->opno);
+	hashedGEOpExpr->opresulttype = get_func_rettype(hashedGEOpExpr->opfuncid);
+	operatorExpressions = lappend(operatorExpressions, hashedGEOpExpr);
+
+	hashedLEOpExpr->opfuncid = get_opcode(hashedLEOpExpr->opno);
+	hashedLEOpExpr->opresulttype = get_func_rettype(hashedLEOpExpr->opfuncid);
+	operatorExpressions = lappend(operatorExpressions, hashedLEOpExpr);
+
+	return operatorExpressions;
+}
+
+
+/*
+ * Returns a Param that can be used as an uninstantiated parameter for the given
+ * column in the sense that paramtype, paramtypmod and collid is set to the input
```

Comment:
> We also use the same code snippet in AddUninstantiatedEqualityQual(). Can we also use this function there?

Oh good catch, I hate duplicate codes. I thought of doing that during the implementation but forgot to do.